### ISSUE TITLE
OneToMany inline fixture definition

### ISSUE THREAD

**Issue by Soullivaneuh** [nelmio/alice#251]
I have the following `TaxRule` entity:

```php
class TaxRule
{
    /**
     * @var int
     *
     * @ORM\Column(name="id", type="integer")
     * @ORM\Id
     * @ORM\GeneratedValue(strategy="AUTO")
     */
    private $id;

    /**
     * @var string
     *
     * @ORM\Column(name="name", type="string", length=64)
     */
    private $name;

    /**
     * @var TaxValue[]|ArrayCollection
     *
     * @ORM\OneToMany(targetEntity="AppBundle\Entity\TaxValue", mappedBy="taxRule", cascade={"persist"}, orphanRemoval=true)
     * @ORM\OrderBy({"beginsAt" = "DESC"})
     */
    private $values;
}
```

`TaxRule` has multiple `TaxValue`, each specific to the rules:

```php
class TaxValue
{
    /**
     * @var int
     *
     * @ORM\Column(name="id", type="integer")
     * @ORM\Id
     * @ORM\GeneratedValue(strategy="AUTO")
     */
    private $id;

    /**
     * @var float
     *
     * @ORM\Column(name="value", type="float")
     */
    private $value;

    /**
     * @var \DateTime
     *
     * @ORM\Column(name="begins_at", type="date")
     */
    private $beginsAt;

    /**
     * @var TaxRule
     *
     * @ORM\ManyToOne(targetEntity="TaxRule", inversedBy="values")
     * @ORM\JoinColumn(name="tax_rule_id", referencedColumnName="id", nullable=false)
     */
    private $taxRule;
}
```

I want to define some tax rule fixtures with them values directly:

```yml
AppBundle\Entity\TaxRule:
    tva_france:
        name: TVA France
        values:
            - { value: 0.2, beginsAt: 2015-01-01 }
    tva_luxembourg:
        name: TVA Luxembourg
        values:
            - { value: 0.15, beginsAt: 2015-01-01 }
```

But this is not working:

```
  [Symfony\Component\Debug\Exception\ContextErrorException]                                                                                                                                                                                                                      
  Catchable Fatal Error: Argument 1 passed to AppBundle\Entity\TaxRule::addValue() must be an instance of AppBundle\Entity\TaxValue, array given, called in /home/sullivan/projects/nexylan/n-admin/vendor/nelmio/alice/src/Nelmio/Alice/Instances/Populator/Methods/ArrayAdd.p  
  hp on line 48 and defined                                                                                                                                                                                                                                                      
```

Is that possible to do that? How? If not, it would be great to have these feature...

Thanks.

**Comment by theofidry** [comment]
You have to create fixtures for `TaxValue` instances and then references them using the `@` annotation. Check out [relation handing#references](https://github.com/nelmio/alice/blob/master/doc/relations-handling.md#references).

**Comment by Soullivaneuh** [comment]
@theofidry I know that, it's especially what I don't want to do.

Each TaxValue will be used for only once TaxRule and information inside is skinny.

Make instance directly on relation would make a clearer fixtures configuration file.

**Comment by theofidry** [comment]
Then the answer is no, at least not to my knowledge. This could be interesting and possible by adding a key value `class`:

```yaml
AppBundle\Entity\TaxRule:
    tva_france:
        name: TVA France
        values:
            - class: AppBundle\Entity\TaxVal
              data: { value: 0.2, beginsAt: 2015-01-01 }
```

**Comment by Soullivaneuh** [comment]
@theofidry I the main idea but I don't think class is necessary. This could be detected on entity definition.

**Comment by cmfcmf** [comment]
I have a similar use case, but on template level. Each *event* has *questions*, whereas **every** *event* has three default questions. I would love to be able to define those inline with the template definition.
```yml
EventManager_Entity_Event:
  base_event (template):
    eventName: <sentence(mt_rand(3, 15))>
    questions (embedded template):
      EventManager_Entity_Question_String:
       lastname (extends lastname_question): {}
        firstname (extends firstname_question): {}
      EventManager_Entity_Question_Email:
        email (extends email_question): {}
```

Now I'd like to be able to just do:
```
EventManager_Entity_Event:
  my_custom_event (extends base_event):
    eventName: Hi there!
```
which then should persist a new event entity which is associated to three new question entities.

**Comment by Soullivaneuh** [comment]
Yes, it would be definitely a nice to have.

**Comment by theofidry** [comment]
@Soullivaneuh btw why you don't want to create a dedicated fixture for it again? For the ease of use?

The only difficulty I may see with that is figuring out how to create you `AppBundle\Entity\TaxVal` entity. One way to do it would be to add a custom `Denormalizer` (which is responsible for transforming the parsed data of your files into a comprehensible set of fixtures) to detect such inlined entities, add a new Fixture with a unique ID which correspond to your inlined object, and replace the value by a reference to this fixture.

But that would be something I would rather put in an alice extension or let the user implement, rather than adding it to the core which is already complex enough. Especially since I already can see some syntax specificities coming which are better being implemented in the user code.

WDYT?

cc @tshelburne

**Comment by tshelburne** [comment]
I agree w/ @theofidry - I think if we can find a middle road for a more friendly "inline" syntax that could be valuable, but I don't think we can make assumptions about reference types b/c we aren't tying the library to Doctrine (or to annotations in general). I'm not opposed to the suggestion @theofidry had, or something similar.

**Comment by tshelburne** [comment]
Or, what about just inline PHP:

```yml
AppBundle\Entity\TaxRule:
    tva_france:
        name: TVA France
        values:
            - <(new AppBundle\Entity\TaxRule(0.2, new Date(2015-01-01))>
    tva_luxembourg:
        name: TVA Luxembourg
        values:
            - <(new AppBundle\Entity\TaxRule(0.15, new Date(2015-01-01))>
```

@theofidry Not sure my syntax is correct.

**Comment by Soullivaneuh** [comment]
Sorry for the very late reply.

The @tshelburne may work, but only if you can pass value on the constructor AFAIK.

Having an inline denormalization would be great. You specify a class on your sample, but I think it should be optional.

We can use a lib such as this one to easy discover the wanted value of the setter/added: https://packagist.org/packages/roave/better-reflection

**Comment by theofidry** [comment]
@Soullivaneuh I think regular reflection can do the job. The point of BetterReflection is to be able to do reflection without loading any code, but in our case we don't care

**Comment by alcalyn** [comment]
I just wonder if this was possible, then found this issue.

I have the same use case, want to do something like:

``` yaml
AppBundle\Entity\TaxRule:
    tva_france:
        name: TVA France
        value:
            coef: 0.2
            beginsAt: 2015-01-01
    tva_luxembourg:
        name: TVA Luxembourg
        value:
            coef: 0.15
            beginsAt: 2015-01-01
```

(in my case this is a `ManyToOne`, not a `OneToMany`)

I know I can create two separate objects, then add a relation with `@tva_france_value`, but I need this just for simplicity purpose : I don't want to create two objects, and two keys (`@tva_france_value` and `@tva_france`). In my case, it's a single relation I wont use elsewhere.

I'm not sure a `class: App\Entity\...` is needed as it could be retrieved from reflection, as it has been mentioned.
